s3metadatarepository/s3metadatarepository.go

```diff
@@ -233,6 +233,76 @@ func (s *S3Repository) Get(ctx context.Context, account, id string) (*dataset.Me
 	return metadata, nil
 }
 
+// Promote sets the finalized_at time and finalized_by user in the metadata object, as well as derivative=false
+func (s *S3Repository) Promote(ctx context.Context, account, id, user string) (*dataset.Metadata, error) {
+	if account == "" {
+		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty account"))
+	}
+
+	if id == "" {
+		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty id"))
+	}
+
+	if user == "" {
+		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty user"))
+	}
+
+	log.Infof("promoting s3metadatarepository '%s' in account '%s'", id, account)
+
+	key := s.Prefix + "/" + account
+	if !strings.HasSuffix(account, "/") && !strings.HasPrefix(id, "/") {
+		key = key + "/"
+	}
+	key = key + id
+
+	out, err := s.S3.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(s.Bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return nil, ErrCode("failed to get metadata object from s3: "+key, err)
+	}
+	defer out.Body.Close()
+
+	metadata := &dataset.Metadata{}
+	err = json.NewDecoder(out.Body).Decode(metadata)
+	if err != nil {
+		return nil, apierror.New(apierror.ErrBadRequest, "failed to decode json from s3", err)
+	}
+
+	log.Debugf("output from getting s3 metadata '%s': %+v", key, metadata)
+
+	if metadata.FinalizedAt != nil {
+		return nil, apierror.New(apierror.ErrConflict, "dataset already finalized", nil)
+	}
+
+	// set the modified/finalized attributes
+	now := time.Now().UTC().Truncate(time.Second)
+	metadata.ModifiedAt = &now
+	metadata.ModifiedBy = user
+	metadata.FinalizedAt = &now
+	metadata.FinalizedBy = user
```

Comment:
should finalizing update the modified time and modified by?  I think it *probably* should but I could argue for both...

Reply:
yeah, i think it makes sense, since it is a modification. I also don't see any reason _not_ to update both